File: kirun-js/__tests__/engine/function/system/date/SetSecondsTest.ts
import { KIRunFunctionRepository, KIRunSchemaRepository, Namespaces } from '../../../../../src';
import { DateFunctionRepository } from '../../../../../src/engine/function/system/date/DateFunctionRepository';
import { GetTimeZoneOffset } from '../../../../../src/engine/function/system/date/GetTimeZoneOffset';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

const dateFunctionRepo = new DateFunctionRepository();

test('testing SetSecondsFunction', async () => {
    let setSeconds = await dateFunctionRepo.find(Namespaces.DATE, 'setSeconds');
    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    );

    if (!setSeconds) {
        throw new Error('Function not available');
    }

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-09-07T17:35:17.123Z'],
            ['secondsValue', -10000],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-09-03T17:35:17.980Z'],
            ['secondsValue', 100],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1970-01-20T15:58:57.561Z'],
            ['secondsValue', 1000],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-10-19T06:44:11.615Z'],
            ['secondsValue', 10000],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '2023-10-24T14:10:30.700+12:00'],
            ['secondsValue', 100],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1994-10-24T14:05:30.406-18:00'],
            ['secondsValue', -100],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1300-10-25T05:42:10.435+14:00'],
            ['secondsValue', -10000],
        ]),
    );

    expect(async () =>
        (await setSeconds?.execute(fep))?.allResults()[0].getResult().get('seconds'),
    ).rejects.toThrowError('Seconds should be in the range of 0 and 59');

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1300-10-25T05:42:10.435+14:00'],
            ['secondsValue', 10],
        ]),
    );

    expect((await setSeconds.execute(fep)).allResults()[0].getResult().get('seconds')).toBe(10);

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1300-10-25T05:42:10.435+14:00'],
            ['secondsValue', 59],
        ]),
    );

    expect((await setSeconds.execute(fep)).allResults()[0].getResult().get('seconds')).toBe(59);

    fep.setArguments(
        new Map<string, any>([
            ['isodate', '1300-10-25T05:42:10.435+14:00'],
            ['secondsValue', 34],
        ]),
    );

    expect((await setSeconds.execute(fep)).allResults()[0].getResult().get('seconds')).toBe(34);
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/SetSecondsTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.namespaces.Namespaces;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class SetSecondsTest {

    DateFunctionRepository dfr = new DateFunctionRepository();

    ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
            new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository());

    String message = "Invalid ISO 8601 Date format.";

    @Test
    void failTest() {

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("abcd"), "secondValue", new JsonPrimitive(121312)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage(message)
                .verify();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("202312=12"), "secondValue", new JsonPrimitive(1245)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectError()
                .verify();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("2053-10-04T14:10:50.700+00:00"), "secondValue",
                new JsonPrimitive(123)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Seconds should be in the range of 0 and 59")
                .verify();

        fep.setArguments(
                Map.of("isoDate", new JsonPrimitive("2023-10-19T23:24:11.615Z"), "secondValue",
                        new JsonPrimitive(-1123)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectErrorMessage("Seconds should be in the range of 0 and 59")
                .verify();

    }

    @Test
    void validTest() {

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("1300-10-25T05:42:13.435+14:00"), "secondValue",
                new JsonPrimitive(10)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("1300-10-25T05:42:10.435+14:00"))
                .verifyComplete();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("1300-10-25T05:42:10.435+14:00"), "secondValue",
                new JsonPrimitive(59)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("1300-10-25T05:42:59.435+14:00"))
                .verifyComplete();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("1300-10-25T05:42:10.435Z"), "secondValue",
                new JsonPrimitive(34)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("1300-10-25T05:42:34.435Z"))
                .verifyComplete();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("-011300-10-25T05:42:10.435Z"), "secondValue",
                new JsonPrimitive(45)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("-011300-10-25T05:42:45.435Z"))
                .verifyComplete();

        fep.setArguments(Map.of("isoDate", new JsonPrimitive("-011300-10-25T05:42:10.435Z"), "secondValue",
                new JsonPrimitive(59)));

        StepVerifier.create(dfr.find(Namespaces.DATE, "SetSeconds")
                .flatMap(e -> e.execute(fep)))
                .expectNextMatches(
                        r -> r.next().getResult().get("result").getAsString().equals("-011300-10-25T05:42:59.435Z"))
                .verifyComplete();
    }

}
